nexus/src/db/db-macros/src/lookup.rs

```diff
@@ -0,0 +1,649 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Procedure macro for generating lookup structures and related functions
+//!
+//! See nexus/src/db/lookup.rs.
+
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+
+//
+// INPUT (arguments to the macro)
+//
+
+/// Arguments for [`lookup_resource!`]
+// NOTE: this is only "pub" for the `cargo doc` link on [`lookup_resource!`].
+#[derive(serde::Deserialize)]
+pub struct Input {
+    /// Name of the resource (PascalCase)
```

Comment:
The "PascalCase" part of this is kinda coincidental - the actual expectation is that:

- `name` is the exact name of a type in `nexus::db::model`
- `name` is the exact name of a type in `nexus::authz`
- The resource exists in `db::schema::<snake_case_name>::dsl`

I tried declaring `lookup_resource` for a resource named `MyNewResource` - I wanted to use "compiler-driven development" just to see where the dependencies are, and it seems like there are quite a handful of these that are implied by the implementation details of the macro.

Reply:
FWIW, I do think documenting the expected behavior is a strong step in the right direction.

I personally find [associated types](https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html) ideal for this sort of association - we're effectively claiming that there exists a resource...
- ... with a model struct
- ... with a schema in the DB
- ... with a corresponding authz resource

but the connection between these pieces of code is implicit.

Like, that's why we're supplying `AuthzKind` by type to the macro, but all the other arguments are strings - it's because they *aren't* types, they must actually correspond to multiple types simultaneously.

To be clear - I don't think this should be blocking - I think the new lookup functions are great, and I love the new API, but I just view this as an area where cleanup could potentially make this more readable by depending on types rather than string-ified tokens.

Reply:
Okay.  I think I'll land this as it is now.  I'm definitely open to cleanup here.  Ideally, we wouldn't need a macro at all to generate all these functions (per our earlier conversation).